### PR TITLE
Implement e2e testing for App Mesh charts

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR adds end-to-end tests for the App Mesh charts:
* create an e2e testing framework based on bash bats
* setup Kubernetes Kind in CircleCI
* install the App Mesh charts and run e2e tests against Kind

Tests:
* assert Helm releases (Helm v2 and v3)
* assert Kubernetes deployments
* assert Kubernetes deployments rollout status
* assert Kubernetes services 

Debug info for failed tests:
* pods status
* pods logs

Successful e2e output:

```
Preparing namespace appmesh-system
namespace/appmesh-system created
customresourcedefinition.apiextensions.k8s.io/meshes.appmesh.k8s.aws created
customresourcedefinition.apiextensions.k8s.io/virtualnodes.appmesh.k8s.aws created
customresourcedefinition.apiextensions.k8s.io/virtualservices.appmesh.k8s.aws created
Installing chart appmesh-controller
✔ appmesh-controller chart install test passed
Waiting for deployment appmesh-controller
deployment "appmesh-controller" successfully rolled out
✔ deployment/appmesh-controller test passed
Installing chart appmesh-inject
✔ appmesh-inject chart install test passed
Waiting for deployment appmesh-inject
deployment "appmesh-inject" successfully rolled out
✔ deployment/appmesh-inject test passed
Waiting for service appmesh-inject
✔ service/appmesh-inject test passed
Installing chart appmesh-prometheus
✔ appmesh-prometheus chart install test passed
Waiting for deployment appmesh-prometheus
deployment "appmesh-prometheus" successfully rolled out
✔ deployment/appmesh-prometheus test passed
Waiting for service appmesh-prometheus
✔ service/appmesh-prometheus test passed
Installing chart appmesh-jaeger
✔ appmesh-jaeger chart install test passed
Waiting for deployment appmesh-jaeger
deployment "appmesh-jaeger" successfully rolled out
✔ deployment/appmesh-jaeger test passed
Waiting for service appmesh-jaeger
✔ service/appmesh-jaeger test passed
```

Failed e2e output:
```
Installing chart appmesh-controller
Waiting for deployment appmesh-controller
deployment "appmesh-controller" successfully rolled out
Installing chart appmesh-inject
Waiting for deployment appmesh-inject
Waiting for deployment "appmesh-inject" rollout to finish: 0 of 1 updated replicas are available...

Listing pods
NAME                                  READY   STATUS             RESTARTS   AGE
appmesh-controller-55c7fc64f5-m6lbp   1/1     Running            0          77s
appmesh-inject-668955fc76-h97lp       0/1     ImagePullBackOff   0          67s

error: timed out waiting for the condition
Error from server (BadRequest): container "appmesh-inject" in pod "appmesh-inject-668955fc76-h97lp" is waiting to start: trying and failing to pull image
```

Fix: #25 

PS. The ECR images can't be used outside AWS (see the above failed test when using ECR) so I had to push the App Mesh images to my Docker Hub account.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/lib/bats.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+if [[ "$1" ]]; then
+  BATS_VERSION=$1
+fi
+
+function installBats() {
+  bin=${REPO_ROOT}/bin
+  mkdir -p $bin
+  cd $bin
+  curl -sSL https://github.com/bats-core/bats-core/archive/v${BATS_VERSION}.tar.gz -o bats.tar.gz
+  tar -xvf bats.tar.gz --strip 1
```
Comment (jaypipes) [inline test/lib/bats.sh]: Maybe do:

```
mv $bin/bin/bats $bin/bats
```
here so that the path to `bats` executable isn't `bin/bin/bats`?

File: stable/appmesh-inject/templates/deployment.yaml [hunk 2]
```diff
@@ -15,6 +15,7 @@ spec:
       labels:
         app.kubernetes.io/name: {{ include "appmesh-inject.name" . }}
         app.kubernetes.io/instance: {{ .Release.Name }}
+        app.kubernetes.io/part-of: appmesh
```
Comment (jaypipes) [inline stable/appmesh-inject/templates/deployment.yaml]: these changes seem unrelated to the e2e test addition. Consider placing these changes into a different PR.

Comment (stefanprodan) [reply stable/appmesh-inject/templates/deployment.yaml]: This is required change for the `teardown` function, having a common label on all App Mesh pods makes it possible to display the logs for debugging.

Comment (jaypipes) [reply stable/appmesh-inject/templates/deployment.yaml]: ack, got it. thanks for the explanation!

File: stable/appmesh-controller/ci/values.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# CI testing values for appmesh-controller.
+
+region: us-west-2
+image:
+  repository: stefanprodan/aws-app-mesh-controller-for-k8s
```
Comment (nckturner) [inline stable/appmesh-controller/ci/values.yaml]: This is temporary, we will replace this once I get the docker repository set up.

File: test/e2e/appmesh.bats [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+export REPO_ROOT=$(git rev-parse --show-toplevel)
+
+namespace=appmesh-system
+charts=${REPO_ROOT}/stable
+
+load ${REPO_ROOT}/test/lib/helm.sh
+
+function setup() {
+  infof "Preparing namespace $namespace"
+  kubectl create ns $namespace >&3
```
Comment (nckturner) [inline test/e2e/appmesh.bats]: Why are we using fd 3 here?

File: test/e2e/appmesh.bats [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+set -o errexit
+
+export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
+export REPO_ROOT=$(git rev-parse --show-toplevel)
+
+namespace=appmesh-system
+charts=${REPO_ROOT}/stable
+
+load ${REPO_ROOT}/test/lib/helm.sh
```
Comment (nckturner) [inline test/e2e/appmesh.bats]: This is handy

File: test/e2e/appmesh.bats (hunk 4 above)
Comment (stefanprodan) [reply test/e2e/appmesh.bats]: This is how you can write to the bats test log.

File: test/lib/bats.sh (hunk 1 above)
Comment (stefanprodan) [reply test/lib/bats.sh]: bats depends on other scripts next to bin, I've moved them all to tmp.

File: test/lib/kind.sh [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  helm init --service-account tiller --upgrade --wait
+
+  echo ">>> Installing Helm $HELM3_VERSION"
+  curl -sSL https://get.helm.sh/helm-${HELM3_VERSION}-linux-amd64.tar.gz | \
+    tar xz && sudo mv linux-amd64/helm /bin/helmv3 && sudo rm -rf linux-amd64
+}
+
+installKubernetes
+
+export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
```
[the remark this answers is not in the document]
Comment (stefanprodan) [reply test/lib/kind.sh]: Fixed by allowing `KUBECONFIG` to be set before running the tests, this way it will work with minikube, eks, etc.

**Comment by stefanprodan** [comment]
@jaypipes @nckturner I think I've addressed all your comments. Thanks for the review!

File: test/lib/kind.sh [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+KIND_VERSION=v0.5.1
+HELM_MODE="${HELM_MODE:-v2}"
+HELM2_VERSION=v2.14.3
+HELM3_VERSION=v3.0.0-beta.5
+
+function installKubernetes() {
+  echo ">>> Installing kubectl"
+  curl -sLO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && \
+  chmod +x kubectl && \
+  sudo mv kubectl /usr/local/bin/
```
Comment (jaypipes) [inline test/lib/kind.sh]: This is not amenable to re-running the `test/lib/kind.sh` script  multiple times, as would be done by a developer testing the repo locally. It will continually download and re-install `kubectl` even if `kubectl` is already installed.

It would be useful to have a `is_installed` helper function that would return 0 when a required binary is installed, 1 otherwise:

Something like this would be good to put into a `test/lib/common.sh` file and source'd here:

```
is_installed() {
    local name="$1"
    if $(which $name >/dev/null 2>&1); then
        return 0
    else
        return 1
    fi
}
```

you can then gate the installation of `kubectl`, `kind` and other executables based on whether they are not already installed.

File: test/lib/kind.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaypipes) [reply test/lib/kind.sh]: What I was asking for was an envvar that would allow me to specify the *kind cluster name* instead of the hard-coded string "kind".

Also, I don't see how the above line won't override a user-supplied `KUBECONFIG` envvar. Wouldn't it need to be this?

```
export KUBECONFIG=${KUBECONFIG:-$(kind get kubeconfig-path --name $KIND_CLUSTER_NAME)}
```

File: test/lib/kind.sh [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+  echo ">>> Installing kind $KIND_VERSION"
+  curl -sSLo kind "https://github.com/kubernetes-sigs/kind/releases/download/${KIND_VERSION}/kind-linux-amd64"
+  chmod +x kind
+  sudo mv kind /usr/local/bin/kind
```
Comment (jaypipes) [inline test/lib/kind.sh]: Same comment about gating with a `is_installed` conditional.

File: test/lib/kind.sh [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  echo ">>> Creating kind cluster"
+  kind create cluster --wait 5m
```
[the remark this answers is not in the document]
Comment (stefanprodan) [reply test/lib/kind.sh]: If you have Kind running locally why would you run the `test/lib/kind.sh` script?

File: test/lib/kind.sh (hunk 7 above)
Comment (stefanprodan) [reply test/lib/kind.sh]: I see no point in running the kind script for local dev. No matter what Kubernetes you may have (Docker for Mac, minikube, Kind, EKS, GKE, etc) you would export the `KUBECONFIG` and run the bats tests.

Comment (jaypipes) [reply test/lib/kind.sh]: I was executing it because you have the Helm installation coupled in the same file and I didn't have Helm installed.
